integration/image_pull_timeout_test.go

```diff
@@ -0,0 +1,459 @@
+
+	t.Run("HoldingContentOpenWriter", testCRIImagePullTimeoutByHoldingContentOpenWriter)
+	t.Run("NoDataTransferred", testCRIImagePullTimeoutByNoDataTransferred)
+}
+
+// testCRIImagePullTimeoutByHoldingContentOpenWriter tests that
+//
+//  It should not cancel if there is no active http requests.
+//
+// When there are several pulling requests for the same blob content, there
+// will only one active http request. It is singleflight. For the waiting pulling
+// request, we should not cancel.
+func testCRIImagePullTimeoutByHoldingContentOpenWriter(t *testing.T) {
+	t.Parallel()
+
+	tmpDir := mkdirTempByTesting(t)
+	defer os.RemoveAll(tmpDir)
+
+	cli := buildLocalContainerdClient(t, tmpDir)
+
+	criService, err := initLocalCRIPlugin(cli, tmpDir, criconfig.Registry{})
+	assert.NoError(t, err)
+
+	ctx := namespaces.WithNamespace(context.Background(), k8sNamespace)
+	contentStore := cli.ContentStore()
+
+	// imageIndexJSON is the manifest of ghcr.io/containerd/registry:2.7.
+	var imageIndexJSON = `
+{
+  "manifests": [
+    {
+      "digest": "sha256:b0b8dd398630cbb819d9a9c2fbd50561370856874b5d5d935be2e0af07c0ff4c",
+      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
+      "platform": {
+        "architecture": "amd64",
+        "os": "linux"
+      },
+      "size": 1363
+    },
+    {
+      "digest": "sha256:6de6b4d5063876c92220d0438ae6068c778d9a2d3845b3d5c57a04a307998df6",
+      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
+      "platform": {
+        "architecture": "arm",
+        "os": "linux",
+        "variant": "v6"
+      },
+      "size": 1363
+    },
+    {
+      "digest": "sha256:c11a277a91045f91866550314a988f937366bc2743859aa0f6ec8ef57b0458ce",
+      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
+      "platform": {
+        "architecture": "arm64",
+        "os": "linux",
+        "variant": "v8"
+      },
+      "size": 1363
+    }
+  ],
+  "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
+  "schemaVersion": 2
+}`
+	var index ocispec.Index
+	assert.NoError(t, json.Unmarshal([]byte(imageIndexJSON), &index))
+
+	var manifestWriters = []io.Closer{}
+
+	cleanupWriters := func() {
+		for _, closer := range manifestWriters {
+			closer.Close()
+		}
+		manifestWriters = manifestWriters[:0]
+	}
+	defer cleanupWriters()
+
+	// hold the writer by the desc
+	for _, desc := range index.Manifests {
+		writer, err := content.OpenWriter(ctx, contentStore,
+			content.WithDescriptor(desc),
+			content.WithRef(fmt.Sprintf("manifest-%v", desc.Digest)),
+		)
+		assert.NoError(t, err, "failed to locked manifest")
+
+		t.Logf("locked the manifest %+v", desc)
+		manifestWriters = append(manifestWriters, writer)
+	}
+
+	errCh := make(chan error)
+	go func() {
+		defer close(errCh)
+
+		_, err := criService.PullImage(ctx, &runtimeapi.PullImageRequest{
+			Image: &runtimeapi.ImageSpec{
+				Image: pullProgressTestImageName,
+			},
+		})
+		errCh <- err
+	}()
```

Comment:
How about using errgroup instead?

Reply:
I need to use select statement with timeout and release the content lock.
If uses errgroup, it looks complicated.